File: app/models/Answer.py
from app.instances import db
from sqlalchemy import select, func
from sqlalchemy.ext.hybrid import hybrid_property
from app.helpers.macros.encode import slugify
import app.models.Language
from app.models.AnswerVote import AnswerVote
from app.models.AnswerRevision import AnswerRevision
from app.helpers.search_index import index_json, IndexStatus, gets_index
import datetime
from math import sqrt
from config import answers

import golflang_encodings


class Answer(db.Model):
    """
    An answer posted to a post by a user.
    """

    __tablename__ = 'answers'
    __table_args__ = {'extend_existing': True}

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    post_id = db.Column(db.Integer, db.ForeignKey('posts.id'), nullable=False)

    language_id = db.Column(db.String(answers['lang_len']), nullable=True, default=None)
    language_name = db.Column(db.String(answers['lang_len']), nullable=True, default=None)

    code = db.Column(db.Text, default=None, nullable=True)
    commentary = db.Column(db.Text, default=None, nullable=True)
    encoding = db.Column(db.String(30), default='UTF-8')
    deleted = db.Column(db.Boolean, nullable=False, default=False)

    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    date_created = db.Column(db.DateTime, default=datetime.datetime.now)

    index_status = db.Column(db.Enum(IndexStatus), default=IndexStatus.UNSYNCHRONIZED, nullable=False)

    user = db.relationship('User', backref='answers')
    post = db.relationship('Post', backref='answers', lazy=True)

    @index_json
    def get_index_json(self):
        last_revision = AnswerRevision.query.filter_by(answer_id=self.id).order_by(AnswerRevision.revision_time.desc()).first()
        if isinstance(last_revision, AnswerRevision):
            last_modified = last_revision.revision_time
        else:
            last_modified = self.date_created

        language = self.get_language()

        return {
            'objectID': f'answer-{self.id}',
            'id': self.id,
            'code': self.code,
            'date_created': self.date_created.isoformat(),
            'last_modified': last_modified.isoformat(),
            'language': language.get_id(),
            'language_name': language.get_display_name(),
            'byte_count': self.byte_len,
            'author': self.user.get_index_json(root_object=False),
            'post': {
                'id': self.post.id,
                'name': self.post.title,
                'slug': slugify(self.post.title)
            }
        }

    def should_index(self):
        return not self.deleted

    @classmethod
    @gets_index
    def get_index(cls):
        return 'answers'

    @classmethod
    def get_index_settings(cls):
        return {
            'searchableAttributes': [
                'code',
                'language_name',
                'author.name',
                'post.name'
            ],
            'attributesToSnippet': [
                'code',
                'language_name',
                'author.name',
                'post.name'
            ]
        }

    @hybrid_property
    def byte_len(self):
        return len(self.code.encode(self.encoding or 'utf8'))

    @byte_len.expression
    def byte_len(cls):
        return func.length(cls.code)

    @hybrid_property
    def score(self):
        ups = sum(vote for vote in self.votes if vote.vote == 1)
        downs = sum(vote for vote in self.votes if vote.vote == -1)

        n = ups + downs

        if n == 0:
            return 0

        z = 1.0
        phat = ups / n
        return (phat + z * z / (2 * n) - z * sqrt((phat * (1 - phat) + z * z / (4 * n)) / n)) / (1 + z * z / n)

    @score.expression
    def score(cls):
        ups = select([func.sum(AnswerVote.vote)]).where(AnswerVote.answer_id == cls.id
                                                        and AnswerVote.vote == 1).label('ups')
        downs = select([func.sum(AnswerVote.vote)]).where(AnswerVote.answer_id == cls.id
                                                          and AnswerVote.vote == -1).label('downs')

        n = ups + downs

        if n == 0:
            return 0

        z = 1.0
        phat = ups / n
        return (phat + z * z / (2 * n) - z * func.sqrt((phat * (1 - phat) + z * z / (4 * n)) / n)) / (1 + z * z / n)

    def to_json(self, no_code=False):
        data = {}

        if not no_code:
            data['code'] = self.code
            data['commentary'] = self.commentary

        data['id'] = self.id
        data['encoding'] = self.encoding
        data['byte_len'] = self.byte_len

        language = self.get_language()
        if language is not None:
            data['lang'] = language.to_json()

        data['owner'] = self.user.to_json()

        data['date_created'] = self.date_created.isoformat()

        data['deleted'] = self.deleted

        return data

    def get_language(self):
        if self.language_id is None:
            return None

        return app.models.Language.Language(self.language_id)

    def revise(self, user, **new_answer_data):
        revision = AnswerRevision(answer_id=self.id,
                                  language_id=self.language_id,
                                  language_name=self.language_name,
                                  code=self.code,
                                  commentary=self.commentary,
                                  encoding=self.encoding,
                                  deleted=self.deleted,
                                  user_id=user.id)

        if 'code' in new_answer_data:
            self.code = new_answer_data['code']

        if 'commentary' in new_answer_data:
            self.commentary = new_answer_data['commentary']

        if 'encoding' in new_answer_data:
            self.encoding = new_answer_data['encoding']

        if 'deleted' in new_answer_data:
            self.deleted = new_answer_data['deleted']

        return self, revision

    def __repr__(self):
        return '<Answer(%r) by %r %s>' % (self.id, self.user.name, "(deleted)" if self.deleted else "")

File: js/models/Answer.js
import ErrorManager from '~/helpers/ErrorManager';
import Language from '~/models/Language';
import Data, { EnvKey } from '~/models/Data';
import User from '~/models/User';
import Post from '~/models/Post';

export const INVALID_JSON = Symbol('User.Error.InvalidJSON');

/**
 * An answer to a challenge
 * @implements {JSONConvertable}
 */
export default class Answer {
    /**
     * Creates an answer from an answer config object
     * @param {Object} answer
     * @param {number} answer.id
     * @param {?string} answer.code
     * @param {?string} answer.encoding
     * @param {boolean} [answer.deleted=false]
     * @param {?string} answer.commentary
     * @param {?number} answer.length
     * @param {?Language} answer.language
     * @param {User} answer.user
     * @param {?Post} answer.post Parent post
     * @param {?Date} answer.dateCreated
     * @param {?Date} answer.dateModified
     */
    constructor({ id, code, encoding, deleted = false, length, language, commentary, user, post, dateCreated, dateModified }) {
        this._id = id;
        this._code = code;
        this._encoding = encoding;
        this._deleted = deleted;
        this._commentary = commentary;
        this._length = length;
        this._language = language;
        this._user = user;
        this._post = post;
        this._dateCreated = dateCreated;
        this._dateModified = dateModified;
    }

    /**
     * Clones and returns a new copy of the answer. This is not deep
     * @return {Answer} a copy
     */
    clone() {
        return new Answer({
            id: this._id,
            code: this._code,
            encoding: this._encoding,
            commentary: this._commentary,
            length: this._length,
            language: this._language,
            user: this._user
        });
    }

    /**
     * @type {?Date}
     */
    get dateCreated() { return this._dateCreated; }

    /**
     * @type {?Date}
     */
    get dateModified() { return this._dateModified; }

    /**
     * @type {?Post}
     */
    get post() { return this._post; }

    /**
     * @type {number}
     */
    get id() { return this._id }

    /**
     * General endpoint for this type of model
     * @type {string}
     */
    get endpoint() { return 'answer' }

    /**
     * Language of ans
     * @type {?Language}
     */
    get language() { return this._language; }

    /**
     * Returns length of answer
     * @type {?number}
     */
    get length() {
        return this._length || (this._code ? this._code.length : null)
    }

    /**
     * Returns the code
     * @type {string}
     */
    get code() { return this._code; }

    /**
     * Sets the code
     * @type {string}
     */
    set code(code) {
        // TODO: support encodings
        this._length = [...code].length;
        this._code = code;
    }

    /**
     * Returns owner of the answer
     * @type {User}
     */
    get user() { return this._user; }

    /**
     * Gets if deleted
     */
    get isDeleted() { return this._deleted; }

    /**
     * Gets if deleted
     */
    set isDeleted(isDeleted) { this._deleted = isDeleted; }

    /**
     * URL of answer
     * @return {string}
     */
    get url() {
        return `${Data.shared.envValueForKey(EnvKey.host)}/answer/${this.id}`;
    }

    /**
     * Converts to json
     * @return {Object} json object
     */
    toJSON() {
        return {
            type: 'answer',
            id: this.id,
            owner: this.user.toJSON(),
            code: this.code,
            deleted: this.isDeleted,
            byte_len: this.length,
            lang: this.language
        };
    }

    /**
     * Unwraps from serach Index JSON object
     * @param {Object} JSON Search index JSON
     * @return {?User} Created object
     */
    static fromIndexJSON(json) {
        return new Answer({
            id: json.id,
            code: json.code,
            deleted: false,
            length: json.byte_count,
            language: Language.fromJSON(json.language),
            user: User.fromIndexJSON(json.author),
            dateCreated: new Date(json.date_created),
            dateModified: new Date(json.last_modified),
            post: new Post({ postId: json.post.id, title: json.post.name, slug: json.post.slug })
        });
    }

    /**
     * Unwraps from an API JSON object.
     * @param {Object} json User JSON object.
     * @return {?Answer} object if succesful, `null` if unauthorized.
     * @throws {TypeError} if invalid JSON object
     */
    static fromJSON(json) {
        // Unwrap all json parameters
        const {
            id,
            code = null,
            encoding = "utf8",
            commentary = "",
            lang: language,
            deleted = false,
            byte_len: length,
            owner
        } = json;

        if (!owner) {
            ErrorManager.raise(`Incomplete Answer JSON`, INVALID_JSON);
        }

        return new Answer({
            id,
            code,
            encoding,
            commentary,
            deleted,
            language: Language.fromJSON(language),
            length,
            user: User.fromJSON(owner)
        });
    }
}
